ethcore/src/client/client.rs

```diff
@@ -554,12 +556,22 @@ impl Client {
 
 		state.journal_under(&mut batch, number, hash).expect("DB commit failed");
 
-		if number >= self.history {
-			let n = number - self.history;
-			if let Some(ancient_hash) = chain.block_hash(n) {
-				state.mark_canonical(&mut batch, n, &ancient_hash).expect("DB commit failed");
-			} else {
-				debug!(target: "client", "Missing expected hash for block {}", n);
+		// prune all ancient eras until we're below the memory target,
+		// but have at least the minimum number of states.
+		loop {
+			if state.journal_db().mem_used() <= self.config.history_mem { break }
```

Comment:
slight concern here about the time taken perform this check potentially multiple times per imported block -- an optimization might be done via incremental bookkeeping using the sizes of key-value pairs inserted and deleted into the pruning overlay rather than sweeping the whole overlay each time.